### PR TITLE
Allow Cross-Origin resource sharing

### PR DESCRIPTION
**Pull request by fr0l**
Allowing Cross-Origin resource sharing makes it possible to use a web-based UI elements inspector

### CODE REVIEW

File: Server/Routes/UndefinedRoutes.m [hunk 1]
```diff
@@ -30,11 +30,17 @@ @implementation UndefinedRoutes
                                      @"requestBody" : body ?: @{}
          }];
     };
+
+    RequestHandler pingBlock = ^(RouteRequest *request, NSDictionary *body, RouteResponse *response) {
```
[the remark this answers is not in the document]
Comment (fr0l) [reply Server/Routes/UndefinedRoutes.m:34]: This is a 'ping/pong' block for to be used with OPTION request. This is required to be able to allow cross origin resource sharing - here is a good detailed documentation https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS#Preflighted_requests

File: Server/Routes/CBXRoute.h [hunk 2]
```diff
@@ -54,5 +54,12 @@ Object containing logic for an HTTP route.
  @param block Block to execude when requests are matched to this route
  */
 + (instancetype)delete:(NSString *)path withBlock:(RequestHandler)block;
+
+/**
+ Convenience constructor for a OPTIONS route
+ @param path Route path regex
+ @param block Block to execude when requests are matched to this route
+ */
++ (instancetype)options:(NSString *)path withBlock:(RequestHandler)block;
```
[the remark this answers is not in the document]
Comment (fr0l) [reply Server/Routes/CBXRoute.h:63]: OPTION request is required as one of three things that need to exist in server-side to be able to allow cross origin resource sharing - here is a good detailed documentation https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS#Preflighted_requests

**Comment by fr0l** [comment]
Hi @jmoody , sorry for not describing the idea in detail. 

in this PR i suggest to allow Cross-Origin resource sharing to allow work of web-page based UI inspector.

One of such Inspectors is this - https://github.com/badoo/DeviceAgent.iOS.Inspector
(it's is a copy of WebDriverAgent's inspector, but it is modified to be able to work with Calabash).

Idea is - you have only two artifacts — `index.html` and `inspector.js` and you could embed them in pretty much any web-server, or just open as is in any web browser.

And from that `index.html` you could connect to any Calabash's DeviceAgent.iOS instance (just modify endpoint) and be able to do quick look for elements.

The only issue with serving the `index.html` from a web browser is any modern browser forbids HTTP requests outside of current domain by default (in particular a request for a screenshot and elements tree).

To allow getting screenshot and elements tree, these requests have to be explicitly allowed on server side — i.e. on HTTP server which is embedded in DeviceAgent.iOS.

Here is detailed description on how the cross-origin resource sharing should work with modern browsers https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS

We have already integrated this solution to [Badoo's fork of DeviceAgent.iOS](https://github.com/badoo/DeviceAgent.iOS/pull/4) , and thought it might be useful to other users of [DeviceAgent.iOS](https://github.com/calabash/DeviceAgent.iOS) project

We plan to enhance the https://github.com/badoo/DeviceAgent.iOS.Inspector a bit to make it more usable.

Hope it helps.
Thank you!

File: Server/CBXCUITestServer.m [hunk 3]
```diff
@@ -37,6 +37,10 @@ - (instancetype)init_private {
         [_server setRouteQueue:dispatch_get_main_queue()];
         [_server setDefaultHeader:@"CalabusDriver"
                                         value:@"CalabashXCUITestServer/1.0"];
+        [_server setDefaultHeader:@"Access-Control-Allow-Origin"
+                                        value:@"*"];
+        [_server setDefaultHeader:@"Access-Control-Allow-Headers"
+                                        value:@"Content-Type, X-Requested-With"];
```
[the remark this answers is not in the document]
Comment (fr0l) [reply Server/CBXCUITestServer.m:43]: please see detailed comment below

File: Server/Routes/CBXRoute.h [hunk 4]
```diff
@@ -54,5 +54,12 @@ Object containing logic for an HTTP route.
  @param block Block to execude when requests are matched to this route
  */
 + (instancetype)delete:(NSString *)path withBlock:(RequestHandler)block;
+
+/**
+ Convenience constructor for a OPTIONS route
```
Comment (BasThomas) [inline Server/Routes/CBXRoute.h:59]:
```suggestion
 Convenience constructor for an OPTIONS route
```

File: Server/Routes/CBXRoute.h [hunk 5]
```diff
@@ -54,5 +54,12 @@ Object containing logic for an HTTP route.
...
+ @param path Route path regex
+ @param block Block to execude when requests are matched to this route
```
Comment (BasThomas) [inline Server/Routes/CBXRoute.h:61]:
```suggestion
 @param block Block to execute when requests are matched to this route
```

(Although I suppose fixing this could go into another pull request as all documentation here seems to include this typo 😄 )

**Comment by jmoody** [comment]
I would appreciate a follow-up pull request.

**Comment by fr0l** [comment]
@BasThomas @jmoody i'll add tests

**Comment by JoeSSS** [comment]
hey, @BasThomas I've read your review wrong.. sorry for merging it too early. @fr0l changelog would also really help. Thanks!

**Comment by BasThomas** [comment]
No worries :)
